src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,514 @@
+            this.Comparer = comparer ?? Comparer<TPriority>.Default;
+        }
+
+        /// <summary>
+        /// Creates a priority queue populated with the specified elements and priorities.
+        /// </summary>
+        public PriorityQueue(IEnumerable<(TElement element, TPriority priority)> items)
+        {
+            if (items is null)
+            {
+                throw new ArgumentNullException(nameof(items));
+            }
+
+            this.nodes = new List<(TElement, TPriority)>(items);
+            this.UnorderedItems = new UnorderedItemsCollection(this.nodes);
+            this.Comparer = Comparer<TPriority>.Default;
+
+            if (this.nodes.Count > 1)
+            {
+                this.Heapify();
+            }
+        }
+
+        /// <summary>
+        /// Creates a priority queue populated with the specified elements and priorities,
+        /// and with the specified priority comparer.
+        /// </summary>
+        public PriorityQueue(IEnumerable<(TElement element, TPriority priority)> items, IComparer<TPriority>? comparer)
+        {
+            if (items is null)
+            {
+                throw new ArgumentNullException(nameof(items));
+            }
+
+            this.nodes = new List<(TElement, TPriority)>(items);
+            this.UnorderedItems = new UnorderedItemsCollection(this.nodes);
+            this.Comparer = comparer ?? Comparer<TPriority>.Default;
+
+            if (this.nodes.Count > 1)
+            {
+                this.Heapify();
+            }
+        }
+
+        /// <summary>
+        /// Gets the current amount of items in the priority queue.
+        /// </summary>
+        public int Count => this.nodes.Count;
+
+        /// <summary>
+        /// Gets the priority comparer of the priority queue.
+        /// </summary>
+        public IComparer<TPriority> Comparer { get; }
+
+        /// <summary>
+        /// Enqueues the specified element and associates it with the specified priority.
+        /// </summary>
+        public void Enqueue(TElement element, TPriority priority)
+        {
+            if (element is null)
+            {
+                throw new ArgumentNullException(nameof(element));
+            }
+
+            if (priority is null)
+            {
+                throw new ArgumentNullException(nameof(priority));
+            }
+
+            // Add the node at the end
+            var node = (element, priority);
+            this.nodes.Add(node);
+
+            // Restore the heap order
+            var lastNodeIndex = this.GetLastNodeIndex();
+            this.MoveUp(node, lastNodeIndex);
+        }
+
+        /// <summary>
+        /// Gets the element associated with the minimal priority.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The queue is empty.</exception>
+        public TElement Peek()
+        {
+            if (this.TryPeek(out var element, out var priority))
```

Comment:
There's no need for all of these `this.`... we generally avoid those in this repo.

Reply:
Ack, will do. Makes sense, given that in this repo there is a convention for using `_` as a prefix for private members (which is what I was aiming at with `.this`, a convention from one of my past teams that I adopted). Will remove all of them. Just out of curiosity — is there any technical disadvantage of using them?

Reply:
>  is there any technical disadvantage of using them?

There's zero impact on the generated IL.